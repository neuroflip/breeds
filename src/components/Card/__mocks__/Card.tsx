const Card = ({ children }: { children: React.JSX.Element}) => {
  return <div data-testid="Card">{ children }</div>
}

export default Card