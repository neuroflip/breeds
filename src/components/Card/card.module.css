.card {
    width: auto;
    max-width: 600px;
    height: 100%;
    margin: 25px;
    border-radius: 15px;
    background-color: var(--tertiary);
    box-shadow: rgba(100, 100, 111, 0.2) 0px 7px 29px 0px;
    padding: 15px;
}

@media (min-width: 768px) {
    .card {
      width: 90%;
    }
}
