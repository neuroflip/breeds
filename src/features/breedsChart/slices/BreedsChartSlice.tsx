import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { BreedsData } from "./types";

const BREAD_API = "https://dog.ceo/api/breeds/list/all"

const getFetchImageAPIUrl = (name: string) => {
  return `https://dog.ceo/api/breed/${name}/images`
}

const initialState: { breeds: BreedsData[]; isLoading: boolean; error: string | null } = {
  breeds: [],
  isLoading: false,
  error: null,
}

export const fetchBreeds = createAsyncThunk('breedsChart/fetchBreeds', async () => {
  const response = await fetch(BREAD_API);
  const data = await response.json();
  const breeds: Array<BreedsData> = Object.keys(data.message).map((breed) => ({ name: breed, images: 0 }))

  const res = breeds.map(async (breed: BreedsData) => {
    const response = await fetch(getFetchImageAPIUrl(breed.name));
    const data = await response.json();
    
    return { name: breed.name, images: data.message.length } ;
  })

  return await Promise.all(res);
});

export const breedsChartSlice = createSlice({
  name: "breedsChart",
  initialState,
  reducers: {
    fetchBreeds(state) {
      state.isLoading = true;
      state.error = null;
    }
  },
  extraReducers: builder => {
    builder
      .addCase(fetchBreeds.pending, (state) => {
        state.isLoading = true
      })
      .addCase(fetchBreeds.fulfilled, (state, action) => {
        state.breeds = action.payload;
        state.isLoading = false;
      })
      .addCase(fetchBreeds.rejected, (state, action) => {
        state.error = action.error.message || "Failed to fetch breeds";
        state.isLoading = false;
      })
  }
});