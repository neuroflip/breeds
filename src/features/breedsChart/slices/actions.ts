export const fetchBreeds = () => {
  return {
    type: `breedsChart/fetchBreeds`
  }
}
