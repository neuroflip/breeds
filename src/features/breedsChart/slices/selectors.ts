import { AppState } from "../../../store/store"
import { BreedsData } from "./types"
import { createSelector } from "@reduxjs/toolkit"
import { shallowEqual } from "react-redux"

export const selectIsLoading = (state: AppState) => {
  return state.breedsChart.isLoading
}

export const selectBreeds = (state: AppState): Array<BreedsData> => {
  return state.breedsChart.breeds
}

const selectTotalImages = (breeds: Array<BreedsData>) => {
  return breeds.reduce((count, breed) => count + breed.value, 0)
}

export const selectBreedsByImagePercentage =  createSelector(
  (state: AppState) => {
    const breeds = state.breedsChart.breeds;
    return { breeds: breeds, totalImages: selectTotalImages(breeds) }
  },
  (state) =>  {
    if (state.totalImages === 0) {
      return [];
    }

    const breedsWithImagePercentage = state.breeds.map((breed) => ({
      name: breed.name, value: Number((breed.value / state.totalImages * 100).toFixed(2))
    }));
  
    return breedsWithImagePercentage.sort((a, b) => b.value - a.value).slice(0, 10);
      
  }, {
    memoizeOptions: {
        resultEqualityCheck: shallowEqual,
    },
}
)

export const selectError = (state: AppState) => {
  return state.breedsChart.error
}