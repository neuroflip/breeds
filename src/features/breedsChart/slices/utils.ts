import { DogApiError, RootThunkAPI } from "./types";

const getFetchImageAPIUrl = (name: string) => {
    return `https://dog.ceo/api/breed/${name}/images`
}

const checkResponseAndThrowErrorIfNeedded = (responseStatus: number, data: DogApiError, thunkApi: RootThunkAPI) =>{
    if (responseStatus !== 200) {
      throw thunkApi.rejectWithValue(data.message);
    }
}

const BREAD_API = "https://dog.ceo/api/breeds/list/all"

export { getFetchImageAPIUrl, checkResponseAndThrowErrorIfNeedded, BREAD_API }
