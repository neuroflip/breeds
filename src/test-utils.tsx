import React, { PropsWithChildren } from 'react'
import { render, renderHook } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import { Provider } from 'react-redux'

import type { AppStore, AppState } from './store/types'
import { setupStore } from './store/store'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<AppState>
  store?: AppStore
}

const renderWithProviders = (ui: React.ReactElement, extendedRenderOptions: ExtendedRenderOptions = {}) => {
  const {
    store = setupStore(),
    ...renderOptions
  } = extendedRenderOptions

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{ children }</Provider>)

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions })
  }
}

const renderHookWithProviders = (render: RenderType, extendedRenderOptions: ExtendedRenderOptions = {}) => {
  const {
    store = setupStore(),
    ...renderOptions
  } = extendedRenderOptions

  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <Provider store={store}>{ children }</Provider>
  );

  return {
    store:extendedRenderOptions.store,
    ...renderHook(render, { wrapper: Wrapper, ...renderOptions }),
  };
};

type RenderType = () => {
  renderTitle: () => React.JSX.Element;
  breedImagesPercent: {
      name: string;
      value: number;
  }[];
}

export { renderWithProviders, renderHookWithProviders }