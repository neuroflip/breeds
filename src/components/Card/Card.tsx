
import { selectError, selectIsLoading } from '../../features/breedsChart/slices/selectors';
import { useAppSelector } from '../../store/hooks';
import ErrorBox from '../ErrorBox/ErrorBox';
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
import styles from './styles/card.module.css';

type CardProps = {
    children: React.ReactNode,
    renderTitle?: () => React.ReactNode,
    customErrorMessage?: string
}

const Card = ({ children, renderTitle, customErrorMessage }: CardProps) => {
    const isLoading = useAppSelector(selectIsLoading)
    const error = useAppSelector(selectError)

    return <article className={styles.card}>
        { renderTitle && renderTitle() }
        { isLoading ? <LoadingSpinner /> :
            error ? <ErrorBox errorMessage={ customErrorMessage ? customErrorMessage : error } /> :
                children }
    </article>
}

export default Card