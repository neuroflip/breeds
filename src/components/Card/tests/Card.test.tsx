import { afterAll, afterEach, beforeAll, describe, expect, test,vi } from 'vitest';
import { act, screen, waitFor } from '@testing-library/react';
import Card from '../Card';
import { configureStore } from '@reduxjs/toolkit';
import { breedsChartSlice, fetchBreeds } from '../../../features/breedsChart/slices/BreedsChartSlice';
import { renderWithProviders } from '../../../test-utils';
import { http, HttpResponse, delay } from 'msw'
import { setupServer, SetupServerApi } from 'msw/node'
import { BREAD_API, getFetchImageAPIUrl } from '../../../features/breedsChart/slices/utils';
import breedsData from './allBreeds.json'
import imagesData from './breedImages.json'

let server: SetupServerApi;
const mockStore = configureStore({
  reducer: {
    breedsChart: breedsChartSlice.reducer
  }
})

vi.mock('../../ErrorBox/ErrorBox')
vi.mock('../../LoadingSpinner/LoadingSpinner')

describe('Card Initial render and Load Content correctly', () => {
  beforeAll(() => {
    const handlers = [
      http.get(BREAD_API, async () => {
        await delay(150)
        return HttpResponse.json(breedsData)
      }),
      http.get(getFetchImageAPIUrl('affenpinscher'), async () => {
          await delay(150)
          return HttpResponse.json(imagesData)
      }),
      http.get(getFetchImageAPIUrl('african'), async () => {
          await delay(150)
          return HttpResponse.json(imagesData)
        })
    ]
    
    server = setupServer(...handlers)
    server.listen()
  })
  afterEach(() => server.resetHandlers())
  afterAll(() => server.close())

  test('initial render before dispatch main action', () => {
    const { container } = renderWithProviders(<Card>
      Card Content
    </Card>, { store: mockStore });

    const cardElement = screen.getByRole('article');

    expect(cardElement).toBeInTheDocument();
    expect(cardElement.tagName).toBe('ARTICLE');
    expect(cardElement.className).toBe('card');
    expect(screen.queryByText('ErrorBox')).not.toBeInTheDocument();
    expect(screen.queryByText('LoadingSpinner')).not.toBeInTheDocument();

    //snapshot
    expect(container).toMatchSnapshot()
  });

  test('initial render before dispatch main action to check title renderProp is called', () => {
    const renderTitle = vi.fn()

    renderWithProviders(<Card renderTitle={renderTitle}>
      Card Content
    </Card>, { store: mockStore });

    const cardElement = screen.getByRole('article');

    expect(cardElement).toBeInTheDocument();
    expect(renderTitle).toHaveBeenCalled();
  });

  test('dispatch main action to load from API and add the card content', async () => {
    renderWithProviders(<Card>
      Card Content
    </Card>, { store: mockStore });
    
    await act(async () =>
      mockStore.dispatch(fetchBreeds())
    )

    waitFor(() => {
      expect(screen.queryByText('ErrorBox')).not.toBeInTheDocument();
      expect(screen.getByText('LoadingSpinner')).toBeInTheDocument();
    })

    const cardElement = screen.getByRole('article');

    expect(cardElement).toBeInTheDocument();
    expect(cardElement.className).toBe('card');
    expect(screen.queryByText('LoadingSpinner')).not.toBeInTheDocument()
    expect(screen.queryByText('ErrorBox')).not.toBeInTheDocument();
    expect(screen.getByText('Card Content')).toBeInTheDocument()
  });
});

describe('Card Load Content correctly with error', () => {
  beforeAll(() => {
    const handlers = [
      http.get(BREAD_API, async () => {
        await delay(150)
        return new HttpResponse(null, { status: 401 })
      })
    ]
    
    server = setupServer(...handlers)
    server.listen()
  })
  afterEach(() => server.resetHandlers())
  afterAll(() => server.close())

  test('dispatch main action and get a load error', async () => {
    renderWithProviders(<Card>
      Card Content
    </Card>, { store: mockStore });
    
    await act(async () =>
      mockStore.dispatch(fetchBreeds())
    )

    waitFor(() => {
      //waitsFor initial loading state
      expect(screen.queryByText('ErrorBox')).not.toBeInTheDocument();
      expect(screen.getByText('LoadingSpinner')).toBeInTheDocument();
    })

    const cardElement = screen.getByRole('article');

    expect(screen.queryByText('LoadingSpinner')).not.toBeInTheDocument()
    expect(screen.queryByText('Card Content')).not.toBeInTheDocument()
    expect(cardElement).toBeInTheDocument();
    expect(cardElement.className).toBe('card');
    expect(screen.getByText('ErrorBox')).toBeInTheDocument()

  })
});