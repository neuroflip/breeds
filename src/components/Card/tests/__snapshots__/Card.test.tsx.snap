// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card Initial render and Load Content correctly > initial render before dispatch main action 1`] = `
<div>
  <article
    class="card"
  >
    Card Content
  </article>
</div>
`;
